092-self-healing-using-cruise-control.md

```diff
@@ -0,0 +1,300 @@
+Cruise Control also have [metrics](https://github.com/linkedin/cruise-control/wiki/Sensors#executor-sensors) for tracking the inter/intra broker replica movement and also the ongoing replica movement rate/concurrency which can be useful to know if all the replicas movement was successful or if there were some replica movement which didn't complete.
+[Work](https://github.com/linkedin/cruise-control/pull/2268) is currently ongoing in the upstream Cruise Control project to have some metrics which can give us the self-healing start and end timestamp based upon the type of anomaly.
+To enable the metrics, the user can follow the Strimzi documentation on how we [expose the metrics](https://strimzi.io/docs/operators/latest/full/deploying.html#assembly-metrics-setup-str) through an HTTP endpoint.
+
+### Proposal for enabling self-healing in Strimzi
+
+In order to activate self-healing, users will need to set the `self.healing.enabled` property to `true` and specifying a notifier that the self-healing mechanism is going to use in the `spec.cruiseControl.config` section of their `Kafka` resource.
+If the user forgets to set the notifier property then the `KubernetesEventNotifier`(custom notifier under Strimzi org) will be used by default and the user will also get a warning in the Kafka CR regarding missing `anomaly.notifier.class` property.
+We will support most anomaly fixes however there are some fixes which we don't want self-healing to fix automatically in a Kubernetes context.
+Broker failure or disk failure fix are disabled by default since we have better ways to fix them instead of moving the partition replicas across the brokers.
+For example, brokers failures are generally due to lack of resources, so we can fix the issue at cluster level and in the same way failed disks can be easily replaced with a new one, so self-healing in these cases can be more resource consuming.
+
+Here is an example of what the configured `Kafka` custom resource could look:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    config:
+      # configuration to enable self-healing for all anomalies
+      self.healing.enabled: true
+      anomaly.notifer.class: <notifier-class> # can be set to `SelfHealingNotifier, SlackSelfHealingNotifier, AlertaSelfHealingNotifier etc. or you own custom notifier
+      self.healing.broker.failure.enabled: false # disables self healing for broker failure
+      self.healing.disk.failure.enabled: false # disables self healing for disk failure
+```
+
+The user can still enable self-healing for disk failures and broker failure if they want by setting the  `self.healing.broker.failure.enabled` for broker failure and `self.healing.disk.failure.enabled` for disk failure to `true`.
```

Comment:
Why? They make zero sense in the Kubernetes context.

I would also expect that if we let the users enable them now, it will be impossible or at least much harder to provide a proper solution for them later as it will need to deal with users who have it enabled.